yasa/hypno.py

```diff
@@ -48,6 +24,919 @@
+        >>> hyp.sleep_statistics()
+        {'TIB': 130.0,
+        'SPT': 110.0,
+        'WASO': 6.0,
+        'TST': 104.0,
+        'SE': 80.0,
+        'SME': 94.5455,
+        'SFI': 1.7308,
+        'SOL': 10.0,
+        'SOL_5min': 15.0,
+        'Lat_REM': 80.0,
+        'WAKE': 26.0,
+        'N1': 4.0,
+        'N2': 45.0,
+        'N3': 30.0,
+        'REM': 25.0,
+        '%N1': 3.8462,
+        '%N2': 43.2692,
+        '%N3': 28.8462,
+        '%REM': 24.0385}
+        """
+        hypno = self.hypno.to_numpy()
+        assert self.n_epochs > 0, "Hypnogram is empty!"
+        all_sleep = ["SLEEP", "N1", "N2", "N3", "NREM", "REM", "LIGHT", "DEEP"]
+        all_non_sleep = ["WAKE", "ART", "UNS"]
+        stats = {}
+
+        # TIB, first and last sleep
+        stats["TIB"] = self.n_epochs
+        idx_sleep = np.where(~np.isin(hypno, all_non_sleep))[0]
+        if not len(idx_sleep):
+            first_sleep, last_sleep = 0, self.n_epochs
+        else:
+            first_sleep = idx_sleep[0]
+            last_sleep = idx_sleep[-1]
+        # Crop to SPT
+        hypno_s = hypno[first_sleep : (last_sleep + 1)]
+        stats["SPT"] = hypno_s.size if len(idx_sleep) else 0
+        stats["WASO"] = hypno_s[hypno_s == "WAKE"].size if len(idx_sleep) else np.nan
+        # Before YASA v0.5.0, TST was calculated as SPT - WASO, meaning that Art
+        # and Unscored epochs were included. TST is now restrained to sleep stages.
+        stats["TST"] = hypno_s[np.isin(hypno_s, all_sleep)].shape[0]
+
+        # Sleep efficiency and fragmentation
+        stats["SE"] = 100 * stats["TST"] / stats["TIB"]
+        if stats["SPT"] == 0:
+            stats["SME"] = np.nan
+            stats["SFI"] = np.nan
+        else:
+            # Sleep maintenance efficiency
+            stats["SME"] = 100 * stats["TST"] / stats["SPT"]
+            # SFI is the ratio of the number of transitions from sleep into Wake to TST (hours)
+            # The original definition included transitions into Wake or N1.
+            counts, _ = self.transition_matrix()
+            n_trans_to_wake = np.sum(
+                counts.loc[
+                    np.intersect1d(counts.index, all_sleep), np.intersect1d(counts.index, ["WAKE"])
+                ].to_numpy()
+            )
+            stats["SFI"] = n_trans_to_wake / (stats["TST"] / (3600 * self.sampling_frequency))
+
+        # Sleep stage latencies -- only relevant if hypno is cropped to TIB
+        stats["SOL"] = first_sleep if stats["TST"] > 0 else np.nan
+        sleep_periods = hypno_find_periods(
+            np.isin(hypno, all_sleep), self.sampling_frequency, threshold="5min"
+        ).query("values == True")
+        if sleep_periods.shape[0]:
+            stats["SOL_5min"] = sleep_periods["start"].iloc[0]
+        else:
+            stats["SOL_5min"] = np.nan
+
+        if "REM" in self.labels:
+            # Question: should we add latencies for other stage too?
```

Comment:
Sorry I missed this before. I suppose it makes sense to go ahead and include them all though I'm indifferent about it.

Reply:
Prefer not to include for now, as the output of this function is already quite dense